underfs/ssh/src/main/java/alluxio/underfs/ssh/SftpFactory.java

```diff
@@ -0,0 +1,351 @@
+ */
+
+package alluxio.underfs.ssh;
+
+import alluxio.AlluxioURI;
+import alluxio.Configuration;
+import alluxio.collections.ConcurrentHashSet;
+import alluxio.underfs.UnderFileSystemConfiguration;
+import alluxio.util.ThreadFactoryUtils;
+
+import com.google.common.base.Preconditions;
+import com.google.common.base.Strings;
+import com.google.common.io.Closer;
+import net.schmizz.sshj.SSHClient;
+import net.schmizz.sshj.sftp.RemoteFile;
+import net.schmizz.sshj.sftp.SFTPClient;
+import net.schmizz.sshj.sftp.SFTPEngine;
+import net.schmizz.sshj.transport.verification.PromiscuousVerifier;
+import net.sf.cglib.proxy.Enhancer;
+import net.sf.cglib.proxy.MethodInterceptor;
+import net.sf.cglib.proxy.MethodProxy;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.lang.reflect.Method;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * The factory used to generate sftp client instance.
+ */
+public class SftpFactory {
+  private static final Logger LOG = LoggerFactory.getLogger(SftpFactory.class);
+  private static final String CLOSE_METHOD = "close";
+  private static final String SSH_GUEST_USER = "guest";
+  private static final int STREAM_CACHE_RW_NUM = 8;
+  private static final long MAX_IDLE_INTERVAL
+      = Configuration.getMs(SshUFSPropertyKey.SSH_MAX_IDLE_INTERVAL);
+
+  private static Set<SftpContainer> sLiveContainers = new ConcurrentHashSet<>();
+  private static ScheduledExecutorService sExecutor = Executors.newSingleThreadScheduledExecutor(
+      ThreadFactoryUtils.build("SftpFactory-close-idle-sftp-%d", true));
+
+  /**
+   * Create a sftp client.
+   * The returned instance is a proxy of SFTPClient in order to support auto connection
+   * and release feature. The ssh and sftp will be connected if the member method called first time,
+   * and the sftp and ssh will be closed if the instance is not used for a while.
+   *
+   * @param uri the {@link AlluxioURI} for this UFS
+   * @param ufsConf UFS configuration
+   * @return the generated instance
+   */
+  public static SFTPClient createSftpClient(AlluxioURI uri, UnderFileSystemConfiguration ufsConf) {
+    SftpContainer container = new SftpContainer(uri, ufsConf);
+    sLiveContainers.add(container);
+    return container.getSftpClientProxy();
+  }
+
+  /**
+   * Transform a remote file to an input stream, which will close the RemoteFile instance
+   * in close method. And the remote file should not be accessed after transformation.
+   *
+   * @param remoteFile the remote file
+   * @param offset the offset
+   * @return the input stream
+   * @throws IOException
+   */
+  public static InputStream transToInputStream(RemoteFile remoteFile, long offset)
```

Comment:
nit: maybe rename to `openAtPosition`?

Reply:
I will rename to createInputStream as next tip